# PyTerrier - Neural Re-Ranking 

Dans ce TP vous allez :

 - reclasser des documents en utilisant des modèles neuronaux comme KNRM, Vanilla BERT, EPIC et monoT5.


# Installation


In [1]:
!pip install -q --upgrade python-terrier

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 104.4/104.4 KB 3.3 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 50.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 69.6/69.6 KB 5.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 311.5/311.5 KB 14.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.5/46.5 KB 2.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 110.5/110.5 KB 8.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 287.2/287.2 KB 13.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.1/45.1 KB 3.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 112.5/112.5 KB 6.0 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 46.2 MB/s eta 

### Installation des plugins Pyterrier  

Nous installons les plugins PyTerrier [OpenNIR](https://opennir.net/) et [monoT5](https://github.com/terrierteam/pyterrier_t5). Vous pouvez ignorer sans risque les erreurs de version des paquets.

In [2]:
!pip install -q --upgrade git+https://github.com/Georgetown-IR-Lab/OpenNIR
!pip install -q --upgrade git+https://github.com/terrierteam/pyterrier_t5

  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.3/114.3 KB 8.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 158.1/158.1 KB 15.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.6/7.6 MB 55.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.8/6.8 MB 111.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 235.9/235.9 KB 28.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 74.6/74.6 KB 10.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 135.6/135.6 KB 3.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 66.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 199.8/199.8 KB 20.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.6/10.6 MB 92.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.6/79.6 KB 11.1 MB/s eta

## Preliminary steps

In [3]:
import os
import tensorflow as tf

**[PyTerrier](https://github.com/terrier-org/pyterrier) initialization** 

In [4]:
import pyterrier as pt
if not pt.started():
    pt.init()

cord19 = pt.datasets.get_dataset('irds:cord19/trec-covid')


from pyterrier.measures import * # allow for natural measure names
import onir_pt

terrier-assemblies 5.7 jar-with-dependencies not found, downloading to /root/.pyterrier...
Done
terrier-python-helper 0.0.7 jar not found, downloading to /root/.pyterrier...
Done


PyTerrier 0.9.2 has loaded Terrier 5.7 (built by craigm on 2022-11-10 18:30) and terrier-helper 0.0.7



Better speed can be achieved with apex installed from https://www.github.com/nvidia/apex.


### Indexation

In [5]:
!rm -rf ./terrier_cord19/*

In [6]:
import os

pt_index_path = './terrier_cord19'

if not os.path.exists(pt_index_path + "/data.properties"):
    # create the index, using the IterDictIndexer indexer 
    indexer = pt.index.IterDictIndexer(pt_index_path, blocks=True)

    # we give the dataset get_corpus_iter() directly to the indexer
    # while specifying the fields to index and the metadata to record
    index_ref = indexer.index(cord19.get_corpus_iter(), 
                              fields=('abstract',), 
                              meta=('docno',))

else:
    # if you already have the index, use it.
    index_ref = pt.IndexRef.of(pt_index_path + "/data.properties")

[INFO] [starting] building docstore
[INFO] If you have a local copy of https://ai2-semanticscholar-cord-19.s3-us-west-2.amazonaws.com/2020-07-16/metadata.csv, you can symlink it here to avoid downloading it again: /root/.ir_datasets/downloads/80d664e496b8b7e50a39c6f6bb92e0ef
[INFO] [starting] https://ai2-semanticscholar-cord-19.s3-us-west-2.amazonaws.com/2020-07-16/metadata.csv
docs_iter:   0%|                                    | 0/192509 s<?, ?doc/s]
https://ai2-semanticscholar-cord-19.s3-us-west-2.amazonaws.com/2020-07-16/metadata.csv: 0.0%| 0.00/269M s<?, ?B/s]
https://ai2-semanticscholar-cord-19.s3-us-west-2.amazonaws.com/2020-07-16/metadata.csv: 0.0%| 65.5k/269M s<10:34, 425kB/s]
https://ai2-semanticscholar-cord-19.s3-us-west-2.amazonaws.com/2020-07-16/metadata.csv: 0.1%| 254k/269M s<05:29, 817kB/s] 
https://ai2-semanticscholar-cord-19.s3-us-west-2.amazonaws.com/2020-07-16/metadata.csv: 0.4%| 950k/269M s<02:13, 2.02MB/s]
https://ai2-semanticscholar-cord-19.s3-us-west-2.amazonaws.

cord19/trec-covid documents:   0%|          | 0/192509 s<?, ?it/s]

<ipython-input-6-59cedea60614>:11: DeprecationWarning: specifying meta and meta_lengths in IterDictIndexer.index() is deprecated, use constructor instead
  index_ref = indexer.index(cord19.get_corpus_iter(),


20:43:31.911 [ForkJoinPool-1-worker-3] WARN org.terrier.structures.indexing.Indexer - Adding an empty document to the index (6iu1dtyl) - further warnings are suppressed
20:44:42.803 [ForkJoinPool-1-worker-3] ERROR org.terrier.structures.indexing.Indexer - Could not finish MetaIndexBuilder: 
java.io.IOException: Key 8lqzfj2e is not unique: 37597,11755
For MetaIndex, to suppress, set metaindex.compressed.reverse.allow.duplicates=true
	at org.terrier.structures.collections.FSOrderedMapFile$MultiFSOMapWriter.mergeTwo(FSOrderedMapFile.java:1374)
	at org.terrier.structures.collections.FSOrderedMapFile$MultiFSOMapWriter.close(FSOrderedMapFile.java:1308)
	at org.terrier.structures.indexing.BaseMetaIndexBuilder.close(BaseMetaIndexBuilder.java:321)
	at org.terrier.structures.indexing.classical.BasicIndexer.indexDocuments(BasicIndexer.java:270)
	at org.terrier.structures.indexing.classical.BasicIndexer.createDirectIndex(BasicIndexer.java:388)
	at org.terrier.structures.indexing.Indexer.index(Inde

## Re-Rankers 

Commençons à explorer quelques méthodes neuronales de re-classement ! Nous pouvons les construire à partir de zéro en utilisant `onir_pt.reranker`.

Le modèle de re-ranking d'OpenNIR est composé de :
 - `ranker` (par exemple, `drmm`, `knrm`, ou `pacrr`). Ceci définit l'architecture neuronale de classement.
 - `vocab` (par exemple, `wordvec_hash`, ou `bert`). Ceci définit comment le texte est encodé par le modèle. Cette approche rend facile l'échange de différentes représentations de texte.

L'exécution de cette ligne prendra quelques minutes car elle télécharge et prépare les vecteurs de mots.

In [7]:
knrm = onir_pt.reranker('knrm', 'wordvec_hash', text_field='abstract')

config file not found: config
[2023-03-31 20:44:55,905][WordvecHashVocab][DEBUG] [starting] downloading https://dl.fbaipublicfiles.com/fasttext/vectors-english/wiki-news-300d-1M.vec.zip


[2023-03-31 20:45:01,714][onir.util.download][WARNING] no hash provided for https://dl.fbaipublicfiles.com/fasttext/vectors-english/wiki-news-300d-1M.vec.zip; consider adding expected_md5="3cc8839ac3fa9a6187149b1e73328b2a" to ensure data integrity.
[2023-03-31 20:45:01,721][onir.util.download][DEBUG] downloaded https://dl.fbaipublicfiles.com/fasttext/vectors-english/wiki-news-300d-1M.vec.zip [5.73s] [682M] [117MB/s]
[2023-03-31 20:45:01,727][WordvecHashVocab][DEBUG] [finished] downloading https://dl.fbaipublicfiles.com/fasttext/vectors-english/wiki-news-300d-1M.vec.zip [5.82s]
[2023-03-31 20:45:01,727][WordvecHashVocab][DEBUG] [starting] extracting vecs
[2023-03-31 20:45:15,388][WordvecHashVocab][DEBUG] [finished] extracting vecs [13.66s]
[2023-03-31 20:45:15,389][WordvecHashVocab][DEBUG] [starting] loading vecs into memory
[2023-03-31 20:47:42,095][WordvecHashVocab][DEBUG] [finished] loading vecs into memory [02:27]
[2023-03-31 20:47:42,345][WordvecHashVocab][DEBUG] [starting] writing

Voyons comment ces modèles fonctionnent pour le classement !

In [8]:
tfidf = pt.BatchRetrieve(index_ref, wmodel="TF_IDF") % 50
get_text = pt.text.get_text(cord19, 'abstract') #>> pt.apply.title_abstract(lambda r: r['title'] + ' ' + r['abstract'])


In [9]:
topics = cord19.get_topics(variant='description')
qrels = cord19.get_qrels()

[INFO] [starting] https://ir.nist.gov/covidSubmit/data/topics-rnd5.xml
[INFO] [finished] https://ir.nist.gov/covidSubmit/data/topics-rnd5.xml: s] [18.7kB] [9.96MB/s]
[INFO] [starting] https://ir.nist.gov/covidSubmit/data/qrels-covid_d5_j0.5-5.txt
[INFO] [finished] https://ir.nist.gov/covidSubmit/data/qrels-covid_d5_j0.5-5.txt: s] [1.14MB] [6.77MB/s]


In [10]:
SEED=42

from sklearn.model_selection import train_test_split

tr_va_topics, test_topics = train_test_split(topics, test_size=15, random_state=SEED)
train_topics, valid_topics =  train_test_split(tr_va_topics, test_size=5, random_state=SEED)


test_qrels = qrels # seulement les annotations des topics en réponse sont utilisés, donc pas de problème si on utilise tout
train_qrels = qrels
valid_qrels = qrels

In [11]:

# build a sub-pipeline to get the concatenated title and abstract text
pipeline = tfidf >> get_text >> knrm
pt.Experiment(
    [tfidf, pipeline],
    test_topics,
    qrels,
    names=['TFIDF', 'TFIDF >> KNRM'],
    eval_metrics=[AP(rel=2), nDCG, nDCG@10, P(rel=2)@10]
)

[2023-03-31 20:48:02,292][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:48:07,375][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/188 s<?, ?it/s]

[2023-03-31 20:48:10,887][onir_pt][DEBUG] [finished] batches: [3.51s] [188it] [53.55it/s]


name  AP(rel=2)      nDCG   nDCG@10  P(rel=2)@10
0          TFIDF   0.054568  0.123589  0.595818     0.546667
1  TFIDF >> KNRM   0.042539  0.112699  0.423870     0.360000

Cela ne fonctionne pas très bien car le modèle n'est pas entraîné ; il utilise des poids aléatoires pour combiner les scores de la matrice de similarité.

## Entraînement du re-ranker

Vous pouvez entraîner des modèles de re-classement dans PyTerrier en utilisant la méthode `fit`. 

In [12]:
pipeline.fit(
    train_topics,
    train_qrels,
    valid_topics,
    valid_qrels)

[2023-03-31 20:48:14,455][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:48:14,457][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:48:14,458][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:48:14,618][onir_pt][DEBUG] [finished] batches: s] [63it] [394.83it/s]
[2023-03-31 20:48:14,734][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:48:14,735][onir_pt][INFO] pre-validation: 0.0053
[2023-03-31 20:48:14,752][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:48:14,752][onir_pt][DEBUG] [starting] training
[2023-03-31 20:48:14,753][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:48:15,018][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:15,530][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:16,639][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:17,850][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:18,335][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:19,442][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:20,708][onir_pt][DEBUG] [finished] train pairs: [5.96s] [1024it] [171.95it/s]
[2023-03-31 20:48:20,712][onir_pt][DEBUG] [finished] training [5.96s]
[2023-03-31 20:48:20,713][onir_pt][INFO] training   it=0 loss=0.2451
[2023-03-31 20:48:20,713][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:48:20,714][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:48:20,715][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:48:20,962][onir_pt][DEBUG] [finished] batches: s] [63it] [259.49it/s]
[2023-03-31 20:48:21,160][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:48:21,163][onir_pt][INFO] validation it=0 map=0.0058 ndcg=0.0151 P_10=0.0700 <--
[2023-03-31 20:48:21,164][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:48:21,165][onir_pt][DEBUG] [starting] training
[2023-03-31 20:48:21,166][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:48:21,532][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:21,639][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:23,492][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:24,457][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:24,733][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:25,131][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:25,258][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:25,737][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:26,222][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:26,399][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:26,537][onir_pt][DEBUG] [finished] train pairs: [5.37s] [1024it] [190.69it/s]
[2023-03-31 20:48:26,540][onir_pt][DEBUG] [finished] training [5.37s]
[2023-03-31 20:48:26,542][onir_pt][INFO] training   it=1 loss=0.2213
[2023-03-31 20:48:26,542][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:48:26,543][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:48:26,544][onir_pt][DEBUG] [starting] 

batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:48:26,710][onir_pt][DEBUG] [finished] batches: s] [63it] [380.44it/s]
[2023-03-31 20:48:26,817][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:48:26,819][onir_pt][INFO] validation it=1 map=0.0059 ndcg=0.0152 P_10=0.0700 <--
[2023-03-31 20:48:26,820][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:48:26,820][onir_pt][DEBUG] [starting] training
[2023-03-31 20:48:26,821][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:48:27,381][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:27,775][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:28,312][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:28,463][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:28,670][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:28,996][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:29,165][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:30,434][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:31,734][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:31,925][onir_pt][DEBUG] [finished] train pairs: [5.10s] [1024it] [200.65it/s]
[2023-03-31 20:48:31,927][onir_pt][DEBUG] [finished] training [5.11s]
[2023-03-31 20:48:31,928][onir_pt][INFO] training   it=2 loss=0.2302
[2023-03-31 20:48:31,928][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:48:31,929][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:48:31,929][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:48:32,090][onir_pt][DEBUG] [finished] batches: s] [63it] [393.18it/s]
[2023-03-31 20:48:32,200][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:48:32,202][onir_pt][INFO] validation it=2 map=0.0059 ndcg=0.0152 P_10=0.0700 <--
[2023-03-31 20:48:32,202][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:48:32,203][onir_pt][DEBUG] [starting] training
[2023-03-31 20:48:32,203][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:48:32,233][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:33,948][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:34,314][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:34,677][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:35,025][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:35,242][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:36,521][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:37,216][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:37,798][onir_pt][DEBUG] [finished] train pairs: [5.59s] [1024it] [183.03it/s]
[2023-03-31 20:48:37,802][onir_pt][DEBUG] [finished] training [5.60s]
[2023-03-31 20:48:37,805][onir_pt][INFO] training   it=3 loss=0.2305
[2023-03-31 20:48:37,806][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:48:37,806][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:48:37,807][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:48:38,069][onir_pt][DEBUG] [finished] batches: s] [63it] [241.31it/s]
[2023-03-31 20:48:38,289][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:48:38,290][onir_pt][INFO] validation it=3 map=0.0059 ndcg=0.0153 P_10=0.0700
[2023-03-31 20:48:38,291][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:48:38,292][onir_pt][DEBUG] [starting] training
[2023-03-31 20:48:38,293][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:48:38,688][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:39,158][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:39,387][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:40,650][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:41,294][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:41,713][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:43,390][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:43,538][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:44,003][onir_pt][DEBUG] [finished] train pairs: [5.71s] [1024it] [179.35it/s]
[2023-03-31 20:48:44,004][onir_pt][DEBUG] [finished] training [5.71s]
[2023-03-31 20:48:44,005][onir_pt][INFO] training   it=4 loss=0.2239
[2023-03-31 20:48:44,008][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:48:44,008][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:48:44,011][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:48:44,169][onir_pt][DEBUG] [finished] batches: s] [63it] [400.87it/s]
[2023-03-31 20:48:44,287][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:48:44,288][onir_pt][INFO] validation it=4 map=0.0058 ndcg=0.0152 P_10=0.0700
[2023-03-31 20:48:44,289][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:48:44,290][onir_pt][DEBUG] [starting] training
[2023-03-31 20:48:44,290][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:48:44,414][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:44,836][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:45,339][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:45,395][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:45,607][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:46,228][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:46,615][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:48,427][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:49,519][onir_pt][DEBUG] [finished] train pairs: [5.23s] [1024it] [195.85it/s]
[2023-03-31 20:48:49,524][onir_pt][DEBUG] [finished] training [5.23s]
[2023-03-31 20:48:49,525][onir_pt][INFO] training   it=5 loss=0.2217
[2023-03-31 20:48:49,526][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:48:49,527][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:48:49,528][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:48:49,689][onir_pt][DEBUG] [finished] batches: s] [63it] [394.33it/s]
[2023-03-31 20:48:49,810][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:48:49,811][onir_pt][INFO] validation it=5 map=0.0058 ndcg=0.0151 P_10=0.0700
[2023-03-31 20:48:49,811][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:48:49,812][onir_pt][DEBUG] [starting] training
[2023-03-31 20:48:49,813][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:48:50,340][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:50,785][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:53,209][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:54,048][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:54,747][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:55,324][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:55,694][onir_pt][DEBUG] [finished] train pairs: [5.88s] [1024it] [174.15it/s]
[2023-03-31 20:48:55,697][onir_pt][DEBUG] [finished] training [5.89s]
[2023-03-31 20:48:55,698][onir_pt][INFO] training   it=6 loss=0.2108
[2023-03-31 20:48:55,699][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:48:55,699][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:48:55,700][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:48:55,941][onir_pt][DEBUG] [finished] batches: s] [63it] [262.74it/s]
[2023-03-31 20:48:56,136][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:48:56,138][onir_pt][INFO] validation it=6 map=0.0059 ndcg=0.0152 P_10=0.0700 <--
[2023-03-31 20:48:56,139][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:48:56,140][onir_pt][DEBUG] [starting] training
[2023-03-31 20:48:56,141][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:48:56,937][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:57,212][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:57,668][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:57,855][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:58,066][onir_pt][DEBUG] not enough negs
[2023-03-31 20:48:59,653][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:00,160][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:01,430][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:01,759][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:01,916][onir_pt][DEBUG] [finished] train pairs: [5.77s] [1024it] [177.34it/s]
[2023-03-31 20:49:01,918][onir_pt][DEBUG] [finished] training [5.78s]
[2023-03-31 20:49:01,919][onir_pt][INFO] training   it=7 loss=0.2140
[2023-03-31 20:49:01,919][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:49:01,920][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:49:01,925][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:49:02,101][onir_pt][DEBUG] [finished] batches: s] [63it] [361.41it/s]
[2023-03-31 20:49:02,216][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:49:02,216][onir_pt][INFO] validation it=7 map=0.0059 ndcg=0.0152 P_10=0.0700
[2023-03-31 20:49:02,217][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:49:02,218][onir_pt][DEBUG] [starting] training
[2023-03-31 20:49:02,219][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:49:02,290][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:03,087][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:03,407][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:03,479][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:03,972][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:05,135][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:05,720][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:05,966][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:06,198][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:06,503][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:07,176][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:07,533][onir_pt][DEBUG] [finished] train pairs: [5.31s] [1024it] [192.72it/s]
[2023-03-31 20:49:07,542][onir_pt][DEBUG] [finished] training [5.32s]
[2023-03-31 20:49:07,542][onir_pt][INFO] training   it=8 loss=0.2278
[2023-03-31 20:49:07,543][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:49:07,544][onir_pt][DEBUG] using GPU (determinis

batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:49:07,713][onir_pt][DEBUG] [finished] batches: s] [63it] [376.41it/s]
[2023-03-31 20:49:07,829][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:49:07,830][onir_pt][INFO] validation it=8 map=0.0059 ndcg=0.0152 P_10=0.0700
[2023-03-31 20:49:07,831][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:49:07,832][onir_pt][DEBUG] [starting] training
[2023-03-31 20:49:07,832][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:49:07,912][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:08,867][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:09,703][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:09,898][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:10,526][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:11,391][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:12,127][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:12,880][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:12,913][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:13,542][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:13,587][onir_pt][DEBUG] [finished] train pairs: [5.75s] [1024it] [177.97it/s]
[2023-03-31 20:49:13,591][onir_pt][DEBUG] [finished] training [5.76s]
[2023-03-31 20:49:13,591][onir_pt][INFO] training   it=9 loss=0.2264
[2023-03-31 20:49:13,592][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:49:13,594][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:49:13,594][onir_pt][DEBUG] [starting] 

batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:49:13,850][onir_pt][DEBUG] [finished] batches: s] [63it] [247.60it/s]
[2023-03-31 20:49:14,081][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:49:14,084][onir_pt][INFO] validation it=9 map=0.0060 ndcg=0.0153 P_10=0.0700 <--
[2023-03-31 20:49:14,085][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:49:14,086][onir_pt][DEBUG] [starting] training
[2023-03-31 20:49:14,087][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:49:15,079][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:15,318][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:15,964][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:16,383][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:16,483][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:17,794][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:17,886][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:18,806][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:19,510][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:20,214][onir_pt][DEBUG] [finished] train pairs: [6.13s] [1024it] [167.18it/s]
[2023-03-31 20:49:20,215][onir_pt][DEBUG] [finished] training [6.13s]
[2023-03-31 20:49:20,218][onir_pt][INFO] training   it=10 loss=0.2233
[2023-03-31 20:49:20,219][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:49:20,219][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:49:20,220][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:49:20,389][onir_pt][DEBUG] [finished] batches: s] [63it] [372.74it/s]
[2023-03-31 20:49:20,510][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:49:20,512][onir_pt][INFO] validation it=10 map=0.0060 ndcg=0.0153 P_10=0.0700 <--
[2023-03-31 20:49:20,513][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:49:20,513][onir_pt][DEBUG] [starting] training
[2023-03-31 20:49:20,514][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:49:21,346][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:21,467][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:21,879][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:22,507][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:23,388][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:24,417][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:25,138][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:25,973][onir_pt][DEBUG] [finished] train pairs: [5.46s] [1024it] [187.60it/s]
[2023-03-31 20:49:25,975][onir_pt][DEBUG] [finished] training [5.46s]
[2023-03-31 20:49:25,976][onir_pt][INFO] training   it=11 loss=0.2149
[2023-03-31 20:49:25,976][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:49:25,977][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:49:25,977][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:49:26,150][onir_pt][DEBUG] [finished] batches: s] [63it] [364.43it/s]
[2023-03-31 20:49:26,274][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:49:26,275][onir_pt][INFO] validation it=11 map=0.0060 ndcg=0.0153 P_10=0.0700
[2023-03-31 20:49:26,276][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:49:26,276][onir_pt][DEBUG] [starting] training
[2023-03-31 20:49:26,277][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:49:26,685][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:26,936][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:29,017][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:29,666][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:29,807][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:30,212][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:30,825][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:31,564][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:32,057][onir_pt][DEBUG] [finished] train pairs: [5.78s] [1024it] [177.16it/s]
[2023-03-31 20:49:32,060][onir_pt][DEBUG] [finished] training [5.78s]
[2023-03-31 20:49:32,061][onir_pt][INFO] training   it=12 loss=0.2161
[2023-03-31 20:49:32,062][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:49:32,062][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:49:32,063][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:49:32,324][onir_pt][DEBUG] [finished] batches: s] [63it] [241.80it/s]
[2023-03-31 20:49:32,549][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:49:32,555][onir_pt][INFO] validation it=12 map=0.0060 ndcg=0.0154 P_10=0.0700 <--
[2023-03-31 20:49:32,556][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:49:32,557][onir_pt][DEBUG] [starting] training
[2023-03-31 20:49:32,558][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:49:32,952][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:33,041][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:34,207][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:34,511][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:36,717][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:37,130][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:37,277][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:37,808][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:38,480][onir_pt][DEBUG] [finished] train pairs: [5.92s] [1024it] [172.94it/s]
[2023-03-31 20:49:38,486][onir_pt][DEBUG] [finished] training [5.93s]
[2023-03-31 20:49:38,487][onir_pt][INFO] training   it=13 loss=0.2355
[2023-03-31 20:49:38,487][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:49:38,488][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:49:38,489][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:49:38,723][onir_pt][DEBUG] [finished] batches: s] [63it] [269.42it/s]
[2023-03-31 20:49:38,920][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:49:38,921][onir_pt][INFO] validation it=13 map=0.0060 ndcg=0.0153 P_10=0.0740
[2023-03-31 20:49:38,921][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:49:38,923][onir_pt][DEBUG] [starting] training
[2023-03-31 20:49:38,923][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:49:39,322][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:39,896][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:40,910][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:42,354][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:42,611][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:44,389][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:44,445][onir_pt][DEBUG] [finished] train pairs: [5.52s] [1024it] [185.51it/s]
[2023-03-31 20:49:44,447][onir_pt][DEBUG] [finished] training [5.52s]
[2023-03-31 20:49:44,447][onir_pt][INFO] training   it=14 loss=0.2230
[2023-03-31 20:49:44,448][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:49:44,448][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:49:44,448][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:49:44,625][onir_pt][DEBUG] [finished] batches: s] [63it] [357.06it/s]
[2023-03-31 20:49:44,733][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:49:44,734][onir_pt][INFO] validation it=14 map=0.0059 ndcg=0.0153 P_10=0.0720
[2023-03-31 20:49:44,735][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:49:44,736][onir_pt][DEBUG] [starting] training
[2023-03-31 20:49:44,736][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:49:44,776][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:45,114][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:46,040][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:46,333][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:46,610][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:48,067][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:49,247][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:49,395][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:49,641][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:50,056][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:50,355][onir_pt][DEBUG] [finished] train pairs: [5.62s] [1024it] [182.25it/s]
[2023-03-31 20:49:50,359][onir_pt][DEBUG] [finished] training [5.62s]
[2023-03-31 20:49:50,360][onir_pt][INFO] training   it=15 loss=0.2189
[2023-03-31 20:49:50,360][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:49:50,361][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:49:50,362][onir_pt][DEBUG] [starting]

batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:49:50,583][onir_pt][DEBUG] [finished] batches: s] [63it] [286.69it/s]
[2023-03-31 20:49:50,801][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:49:50,808][onir_pt][INFO] validation it=15 map=0.0060 ndcg=0.0153 P_10=0.0720
[2023-03-31 20:49:50,809][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:49:50,810][onir_pt][DEBUG] [starting] training
[2023-03-31 20:49:50,810][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:49:51,272][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:51,784][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:51,882][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:52,023][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:52,251][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:52,655][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:55,666][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:55,762][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:56,945][onir_pt][DEBUG] [finished] train pairs: [6.13s] [1024it] [166.94it/s]
[2023-03-31 20:49:56,949][onir_pt][DEBUG] [finished] training [6.14s]
[2023-03-31 20:49:56,950][onir_pt][INFO] training   it=16 loss=0.2246
[2023-03-31 20:49:56,950][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:49:56,951][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:49:56,952][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:49:57,230][onir_pt][DEBUG] [finished] batches: s] [63it] [227.60it/s]
[2023-03-31 20:49:57,468][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:49:57,469][onir_pt][INFO] validation it=16 map=0.0060 ndcg=0.0154 P_10=0.0740
[2023-03-31 20:49:57,474][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:49:57,475][onir_pt][DEBUG] [starting] training
[2023-03-31 20:49:57,476][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:49:57,677][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:58,549][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:58,787][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:59,362][onir_pt][DEBUG] not enough negs
[2023-03-31 20:49:59,489][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:00,091][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:00,240][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:01,192][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:01,526][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:01,630][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:02,248][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:03,047][onir_pt][DEBUG] [finished] train pairs: [5.57s] [1024it] [183.82it/s]
[2023-03-31 20:50:03,051][onir_pt][DEBUG] [finished] training [5.58s]
[2023-03-31 20:50:03,052][onir_pt][INFO] training   it=17 loss=0.2331
[2023-03-31 20:50:03,052][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:50:03,053][onir_pt][DEBUG] using GPU (determini

batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:50:03,289][onir_pt][DEBUG] [finished] batches: s] [63it] [268.86it/s]
[2023-03-31 20:50:03,495][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:50:03,498][onir_pt][INFO] validation it=17 map=0.0061 ndcg=0.0155 P_10=0.0760 <--
[2023-03-31 20:50:03,499][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:50:03,499][onir_pt][DEBUG] [starting] training
[2023-03-31 20:50:03,500][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:50:03,543][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:03,631][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:04,031][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:04,250][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:04,755][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:05,690][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:06,057][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:06,690][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:08,372][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:08,860][onir_pt][DEBUG] [finished] train pairs: [5.36s] [1024it] [191.07it/s]
[2023-03-31 20:50:08,864][onir_pt][DEBUG] [finished] training [5.36s]
[2023-03-31 20:50:08,864][onir_pt][INFO] training   it=18 loss=0.2158
[2023-03-31 20:50:08,865][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:50:08,866][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:50:08,867][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:50:09,153][onir_pt][DEBUG] [finished] batches: s] [63it] [223.85it/s]
[2023-03-31 20:50:09,388][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:50:09,389][onir_pt][INFO] validation it=18 map=0.0059 ndcg=0.0152 P_10=0.0700
[2023-03-31 20:50:09,390][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:50:09,391][onir_pt][DEBUG] [starting] training
[2023-03-31 20:50:09,392][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:50:10,125][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:10,427][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:12,181][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:13,530][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:13,767][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:14,010][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:14,267][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:15,579][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:15,637][onir_pt][DEBUG] [finished] train pairs: [6.24s] [1024it] [163.97it/s]
[2023-03-31 20:50:15,640][onir_pt][DEBUG] [finished] training [6.25s]
[2023-03-31 20:50:15,642][onir_pt][INFO] training   it=19 loss=0.2302
[2023-03-31 20:50:15,642][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:50:15,642][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:50:15,643][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:50:15,913][onir_pt][DEBUG] [finished] batches: s] [63it] [233.16it/s]
[2023-03-31 20:50:16,177][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:50:16,178][onir_pt][INFO] validation it=19 map=0.0059 ndcg=0.0152 P_10=0.0700
[2023-03-31 20:50:16,179][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:50:16,180][onir_pt][DEBUG] [starting] training
[2023-03-31 20:50:16,181][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:50:16,661][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:16,739][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:16,903][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:17,129][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:18,936][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:19,632][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:20,089][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:20,868][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:21,705][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:21,888][onir_pt][DEBUG] [finished] train pairs: [5.71s] [1024it] [179.44it/s]
[2023-03-31 20:50:21,890][onir_pt][DEBUG] [finished] training [5.71s]
[2023-03-31 20:50:21,891][onir_pt][INFO] training   it=20 loss=0.2043
[2023-03-31 20:50:21,893][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:50:21,893][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:50:21,894][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:50:22,146][onir_pt][DEBUG] [finished] batches: s] [63it] [249.76it/s]
[2023-03-31 20:50:22,362][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:50:22,363][onir_pt][INFO] validation it=20 map=0.0060 ndcg=0.0153 P_10=0.0700
[2023-03-31 20:50:22,363][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:50:22,365][onir_pt][DEBUG] [starting] training
[2023-03-31 20:50:22,365][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:50:22,662][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:23,278][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:23,450][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:23,794][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:25,091][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:25,419][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:26,276][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:26,790][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:27,081][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:28,317][onir_pt][DEBUG] [finished] train pairs: [5.95s] [1024it] [172.08it/s]
[2023-03-31 20:50:28,320][onir_pt][DEBUG] [finished] training [5.96s]
[2023-03-31 20:50:28,321][onir_pt][INFO] training   it=21 loss=0.2261
[2023-03-31 20:50:28,321][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:50:28,322][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:50:28,323][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:50:28,545][onir_pt][DEBUG] [finished] batches: s] [63it] [284.13it/s]
[2023-03-31 20:50:28,741][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:50:28,742][onir_pt][INFO] validation it=21 map=0.0059 ndcg=0.0152 P_10=0.0700
[2023-03-31 20:50:28,742][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:50:28,743][onir_pt][DEBUG] [starting] training
[2023-03-31 20:50:28,744][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:50:28,987][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:29,558][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:29,826][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:30,361][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:31,497][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:32,370][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:32,792][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:33,159][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:33,165][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:34,460][onir_pt][DEBUG] [finished] train pairs: [5.72s] [1024it] [179.18it/s]
[2023-03-31 20:50:34,466][onir_pt][DEBUG] [finished] training [5.72s]
[2023-03-31 20:50:34,467][onir_pt][INFO] training   it=22 loss=0.2248
[2023-03-31 20:50:34,467][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:50:34,468][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:50:34,469][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:50:34,751][onir_pt][DEBUG] [finished] batches: s] [63it] [223.84it/s]
[2023-03-31 20:50:35,016][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:50:35,017][onir_pt][INFO] validation it=22 map=0.0059 ndcg=0.0152 P_10=0.0700
[2023-03-31 20:50:35,018][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:50:35,019][onir_pt][DEBUG] [starting] training
[2023-03-31 20:50:35,019][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:50:35,624][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:35,968][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:36,264][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:36,942][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:37,418][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:39,930][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:40,568][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:40,892][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:40,902][onir_pt][DEBUG] [finished] train pairs: [5.88s] [1024it] [174.10it/s]
[2023-03-31 20:50:40,905][onir_pt][DEBUG] [finished] training [5.89s]
[2023-03-31 20:50:40,906][onir_pt][INFO] training   it=23 loss=0.2214
[2023-03-31 20:50:40,906][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:50:40,907][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:50:40,908][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:50:41,172][onir_pt][DEBUG] [finished] batches: s] [63it] [239.45it/s]
[2023-03-31 20:50:41,385][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:50:41,386][onir_pt][INFO] validation it=23 map=0.0059 ndcg=0.0152 P_10=0.0700
[2023-03-31 20:50:41,392][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:50:41,393][onir_pt][DEBUG] [starting] training
[2023-03-31 20:50:41,394][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:50:42,044][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:42,161][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:42,914][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:43,994][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:45,137][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:45,696][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:47,061][onir_pt][DEBUG] [finished] train pairs: [5.67s] [1024it] [180.70it/s]
[2023-03-31 20:50:47,065][onir_pt][DEBUG] [finished] training [5.67s]
[2023-03-31 20:50:47,065][onir_pt][INFO] training   it=24 loss=0.2259
[2023-03-31 20:50:47,066][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:50:47,068][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:50:47,071][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:50:47,327][onir_pt][DEBUG] [finished] batches: s] [63it] [246.91it/s]
[2023-03-31 20:50:47,588][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:50:47,594][onir_pt][INFO] validation it=24 map=0.0059 ndcg=0.0153 P_10=0.0700
[2023-03-31 20:50:47,595][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:50:47,596][onir_pt][DEBUG] [starting] training
[2023-03-31 20:50:47,596][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:50:47,908][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:48,465][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:49,201][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:49,818][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:50,811][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:51,268][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:51,722][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:52,136][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:53,413][onir_pt][DEBUG] [finished] train pairs: [5.82s] [1024it] [176.06it/s]
[2023-03-31 20:50:53,418][onir_pt][DEBUG] [finished] training [5.82s]
[2023-03-31 20:50:53,419][onir_pt][INFO] training   it=25 loss=0.2155
[2023-03-31 20:50:53,420][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:50:53,420][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:50:53,421][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:50:53,693][onir_pt][DEBUG] [finished] batches: s] [63it] [232.96it/s]
[2023-03-31 20:50:53,939][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:50:53,940][onir_pt][INFO] validation it=25 map=0.0060 ndcg=0.0153 P_10=0.0720
[2023-03-31 20:50:53,941][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:50:53,942][onir_pt][DEBUG] [starting] training
[2023-03-31 20:50:53,945][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:50:54,475][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:55,616][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:56,292][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:57,050][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:58,566][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:59,231][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:59,728][onir_pt][DEBUG] not enough negs
[2023-03-31 20:50:59,837][onir_pt][DEBUG] [finished] train pairs: [5.89s] [1024it] [173.83it/s]
[2023-03-31 20:50:59,840][onir_pt][DEBUG] [finished] training [5.90s]
[2023-03-31 20:50:59,841][onir_pt][INFO] training   it=26 loss=0.2215
[2023-03-31 20:50:59,842][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:50:59,842][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:50:59,843][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:51:00,052][onir_pt][DEBUG] [finished] batches: s] [63it] [303.25it/s]
[2023-03-31 20:51:00,250][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:51:00,251][onir_pt][INFO] validation it=26 map=0.0060 ndcg=0.0153 P_10=0.0740
[2023-03-31 20:51:00,252][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:51:00,253][onir_pt][DEBUG] [starting] training
[2023-03-31 20:51:00,254][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:51:00,407][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:00,884][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:02,106][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:03,188][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:04,058][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:04,539][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:05,543][onir_pt][DEBUG] [finished] train pairs: [5.29s] [1024it] [193.66it/s]
[2023-03-31 20:51:05,547][onir_pt][DEBUG] [finished] training [5.29s]
[2023-03-31 20:51:05,548][onir_pt][INFO] training   it=27 loss=0.2191
[2023-03-31 20:51:05,548][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:51:05,549][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:51:05,550][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:51:05,768][onir_pt][DEBUG] [finished] batches: s] [63it] [290.02it/s]
[2023-03-31 20:51:05,991][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:51:05,992][onir_pt][INFO] validation it=27 map=0.0061 ndcg=0.0155 P_10=0.0740
[2023-03-31 20:51:05,993][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:51:05,994][onir_pt][DEBUG] [starting] training
[2023-03-31 20:51:05,995][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:51:06,823][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:06,980][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:07,182][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:07,553][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:08,111][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:08,907][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:09,685][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:09,851][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:10,311][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:10,763][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:10,906][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:11,866][onir_pt][DEBUG] [finished] train pairs: [5.87s] [1024it] [174.42it/s]
[2023-03-31 20:51:11,870][onir_pt][DEBUG] [finished] training [5.88s]
[2023-03-31 20:51:11,871][onir_pt][INFO] training   it=28 loss=0.2141
[2023-03-31 20:51:11,871][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:51:11,872][onir_pt][DEBUG] using GPU (determini

batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:51:12,135][onir_pt][DEBUG] [finished] batches: s] [63it] [240.60it/s]
[2023-03-31 20:51:12,375][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:51:12,378][onir_pt][INFO] validation it=28 map=0.0060 ndcg=0.0154 P_10=0.0740
[2023-03-31 20:51:12,378][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:51:12,379][onir_pt][DEBUG] [starting] training
[2023-03-31 20:51:12,380][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:51:12,608][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:12,855][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:13,081][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:13,351][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:13,640][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:13,992][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:14,088][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:15,982][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:16,224][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:16,539][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:17,012][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:17,606][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:18,442][onir_pt][DEBUG] [finished] train pairs: [6.06s] [1024it] [168.95it/s]
[2023-03-31 20:51:18,446][onir_pt][DEBUG] [finished] training [6.07s]
[2023-03-31 20:51:18,446][onir_pt][INFO] training   it=29 loss=0.2296
[2023-03-31 20:51:18,450][onir_pt][DEBUG] [starting] validation
[202

batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:51:18,679][onir_pt][DEBUG] [finished] batches: s] [63it] [278.31it/s]
[2023-03-31 20:51:18,911][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:51:18,913][onir_pt][INFO] validation it=29 map=0.0060 ndcg=0.0153 P_10=0.0720
[2023-03-31 20:51:18,914][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:51:18,915][onir_pt][DEBUG] [starting] training
[2023-03-31 20:51:18,915][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:51:19,733][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:19,916][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:20,343][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:20,405][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:20,788][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:21,471][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:22,550][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:22,802][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:23,045][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:23,908][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:24,105][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:24,863][onir_pt][DEBUG] [finished] train pairs: [5.95s] [1024it] [172.19it/s]
[2023-03-31 20:51:24,867][onir_pt][DEBUG] [finished] training [5.95s]
[2023-03-31 20:51:24,867][onir_pt][INFO] training   it=30 loss=0.2229
[2023-03-31 20:51:24,868][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:51:24,872][onir_pt][DEBUG] using GPU (determini

batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:51:25,134][onir_pt][DEBUG] [finished] batches: s] [63it] [242.62it/s]
[2023-03-31 20:51:25,380][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:51:25,382][onir_pt][INFO] validation it=30 map=0.0060 ndcg=0.0153 P_10=0.0720
[2023-03-31 20:51:25,386][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:51:25,387][onir_pt][DEBUG] [starting] training
[2023-03-31 20:51:25,388][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:51:26,358][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:27,071][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:27,208][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:27,409][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:28,869][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:29,263][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:30,164][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:30,947][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:31,226][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:31,413][onir_pt][DEBUG] [finished] train pairs: [6.02s] [1024it] [169.97it/s]
[2023-03-31 20:51:31,416][onir_pt][DEBUG] [finished] training [6.03s]
[2023-03-31 20:51:31,417][onir_pt][INFO] training   it=31 loss=0.2253
[2023-03-31 20:51:31,418][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:51:31,419][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:51:31,423][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:51:31,662][onir_pt][DEBUG] [finished] batches: s] [63it] [264.06it/s]
[2023-03-31 20:51:31,854][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:51:31,854][onir_pt][INFO] validation it=31 map=0.0061 ndcg=0.0155 P_10=0.0740
[2023-03-31 20:51:31,855][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:51:31,856][onir_pt][DEBUG] [starting] training
[2023-03-31 20:51:31,857][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:51:32,146][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:32,228][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:32,688][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:32,989][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:35,770][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:36,101][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:36,500][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:36,699][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:37,115][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:37,343][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:37,463][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:37,733][onir_pt][DEBUG] [finished] train pairs: [5.88s] [1024it] [174.27it/s]
[2023-03-31 20:51:37,737][onir_pt][DEBUG] [finished] training [5.88s]
[2023-03-31 20:51:37,738][onir_pt][INFO] training   it=32 loss=0.2227
[2023-03-31 20:51:37,738][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:51:37,739][onir_pt][DEBUG] using GPU (determini

batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:51:37,960][onir_pt][DEBUG] [finished] batches: s] [63it] [286.74it/s]
[2023-03-31 20:51:38,156][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:51:38,157][onir_pt][INFO] validation it=32 map=0.0060 ndcg=0.0153 P_10=0.0720
[2023-03-31 20:51:38,157][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:51:38,158][onir_pt][DEBUG] [starting] training
[2023-03-31 20:51:38,159][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:51:39,806][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:40,045][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:42,831][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:43,371][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:43,610][onir_pt][DEBUG] [finished] train pairs: [5.45s] [1024it] [187.87it/s]
[2023-03-31 20:51:43,614][onir_pt][DEBUG] [finished] training [5.46s]
[2023-03-31 20:51:43,614][onir_pt][INFO] training   it=33 loss=0.2176
[2023-03-31 20:51:43,615][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:51:43,616][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:51:43,617][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:51:43,849][onir_pt][DEBUG] [finished] batches: s] [63it] [272.55it/s]
[2023-03-31 20:51:44,046][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:51:44,047][onir_pt][INFO] validation it=33 map=0.0060 ndcg=0.0153 P_10=0.0700
[2023-03-31 20:51:44,047][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:51:44,048][onir_pt][DEBUG] [starting] training
[2023-03-31 20:51:44,049][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:51:45,690][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:45,949][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:46,290][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:46,737][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:46,992][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:47,090][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:47,661][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:48,241][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:48,391][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:48,535][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:49,035][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:49,193][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:50,023][onir_pt][DEBUG] [finished] train pairs: [5.97s] [1024it] [171.45it/s]
[2023-03-31 20:51:50,027][onir_pt][DEBUG] [finished] training [5.98s]
[2023-03-31 20:51:50,027][onir_pt][INFO] training   it=34 loss=0.2335
[2023-03-31 20:51:50,028][onir_pt][DEBUG] [starting] validation
[202

batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:51:50,297][onir_pt][DEBUG] [finished] batches: s] [63it] [237.00it/s]
[2023-03-31 20:51:50,567][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:51:50,569][onir_pt][INFO] validation it=34 map=0.0059 ndcg=0.0152 P_10=0.0700
[2023-03-31 20:51:50,569][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:51:50,571][onir_pt][DEBUG] [starting] training
[2023-03-31 20:51:50,572][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:51:51,290][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:51,640][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:52,279][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:52,963][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:53,189][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:53,670][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:54,044][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:54,791][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:55,910][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:55,964][onir_pt][DEBUG] [finished] train pairs: [5.39s] [1024it] [189.94it/s]
[2023-03-31 20:51:55,968][onir_pt][DEBUG] [finished] training [5.40s]
[2023-03-31 20:51:55,968][onir_pt][INFO] training   it=35 loss=0.2311
[2023-03-31 20:51:55,969][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:51:55,970][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:51:55,970][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:51:56,175][onir_pt][DEBUG] [finished] batches: s] [63it] [308.14it/s]
[2023-03-31 20:51:56,397][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:51:56,398][onir_pt][INFO] validation it=35 map=0.0059 ndcg=0.0153 P_10=0.0700
[2023-03-31 20:51:56,399][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:51:56,400][onir_pt][DEBUG] [starting] training
[2023-03-31 20:51:56,400][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:51:56,664][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:56,924][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:57,096][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:57,316][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:57,569][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:57,743][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:58,984][onir_pt][DEBUG] not enough negs
[2023-03-31 20:51:59,706][onir_pt][DEBUG] not enough negs
[2023-03-31 20:52:01,509][onir_pt][DEBUG] not enough negs
[2023-03-31 20:52:01,539][onir_pt][DEBUG] [finished] train pairs: [5.14s] [1024it] [199.31it/s]
[2023-03-31 20:52:01,542][onir_pt][DEBUG] [finished] training [5.14s]
[2023-03-31 20:52:01,543][onir_pt][INFO] training   it=36 loss=0.2278
[2023-03-31 20:52:01,544][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:52:01,544][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:52:01,545][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:52:01,757][onir_pt][DEBUG] [finished] batches: s] [63it] [297.98it/s]
[2023-03-31 20:52:01,960][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:52:01,962][onir_pt][INFO] validation it=36 map=0.0060 ndcg=0.0153 P_10=0.0720
[2023-03-31 20:52:01,963][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:52:01,964][onir_pt][DEBUG] [starting] training
[2023-03-31 20:52:01,964][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2023-03-31 20:52:02,233][onir_pt][DEBUG] not enough negs
[2023-03-31 20:52:03,410][onir_pt][DEBUG] not enough negs
[2023-03-31 20:52:04,619][onir_pt][DEBUG] not enough negs
[2023-03-31 20:52:05,355][onir_pt][DEBUG] not enough negs
[2023-03-31 20:52:05,498][onir_pt][DEBUG] not enough negs
[2023-03-31 20:52:06,174][onir_pt][DEBUG] not enough negs
[2023-03-31 20:52:06,226][onir_pt][DEBUG] not enough negs
[2023-03-31 20:52:07,083][onir_pt][DEBUG] not enough negs
[2023-03-31 20:52:07,362][onir_pt][DEBUG] [finished] train pairs: [5.40s] [1024it] [189.73it/s]
[2023-03-31 20:52:07,366][onir_pt][DEBUG] [finished] training [5.40s]
[2023-03-31 20:52:07,367][onir_pt][INFO] training   it=37 loss=0.2167
[2023-03-31 20:52:07,368][onir_pt][DEBUG] [starting] validation
[2023-03-31 20:52:07,368][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:52:07,369][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2023-03-31 20:52:07,616][onir_pt][DEBUG] [finished] batches: s] [63it] [255.91it/s]
[2023-03-31 20:52:07,893][onir_pt][DEBUG] [finished] validation s]
[2023-03-31 20:52:07,895][onir_pt][INFO] validation it=37 map=0.0059 ndcg=0.0152 P_10=0.0700
[2023-03-31 20:52:07,895][onir_pt][INFO] early stopping; model reverting back to it=17


In [13]:
pt.Experiment(
    [tfidf, pipeline],
    test_topics,
    qrels,
    names=['TFIDF', 'TFIDF >> KNRM (trained)'],
    eval_metrics=[AP(rel=2), nDCG, nDCG@10, P(rel=2)@10]
)

[2023-03-31 20:52:10,966][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:52:10,968][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/188 s<?, ?it/s]

[2023-03-31 20:52:11,695][onir_pt][DEBUG] [finished] batches: s] [188it] [259.01it/s]


name  AP(rel=2)      nDCG   nDCG@10  P(rel=2)@10
0                    TFIDF   0.054568  0.123589  0.595818     0.546667
1  TFIDF >> KNRM (trained)   0.051552  0.121711  0.506577     0.420000

In [26]:
#je n'arrive pas a diminuer la taille des donnees, j'ai essaye plusieurs methodes mais aucune n'a fonctionne. Je ne peux pas tester le code sur google colab 

Le résultat est plus élévé, mais cela reste moins bon que le TFIDF. Proposez une hypothèse sur le problème.

## Vanilla BERT

Les modèles linguistiques contextualisés, tels que [BERT] (https://arxiv.org/abs/1810.04805), sont des modèles neuronaux beaucoup plus puissants qui se sont avérés efficaces pour le classement.

Nous allons essayer d'utiliser une version "vanille" (ou "mono") du modèle BERT. Le modèle BERT est pré-entraîné pour la modélisation du langage et la prédiction de la phrase suivante.

In [14]:
#del knrm # clear out memory from KNRM
vbert = onir_pt.reranker('vanilla_transformer', 'bert', text_field='abstract', vocab_config={'train': True})

100%|██████████| 231508/231508 s<0ms, 22230704.24B/s]
100%|██████████| 433/433 s<0ms, 280917.81B/s]
100%|██████████| 440473133/440473133 [9.97s<0ms, 44191589.33B/s] 


Voyons comment ce modèle se comporte sur la collection TREC COVID.

In [15]:
pipeline = tfidf % 50 >> get_text >> vbert
pt.Experiment(
    [tfidf, pipeline],
    test_topics,
    qrels,
    names=['TFIDF', 'TFIDF >> VBERT'],
    baseline=0,
    eval_metrics=[AP(rel=2), nDCG, nDCG@10, P(rel=2)@10]
)

[2023-03-31 20:52:35,540][onir_pt][DEBUG] using GPU (deterministic)
[2023-03-31 20:52:35,696][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/188 s<?, ?it/s]

[2023-03-31 20:52:56,451][onir_pt][DEBUG] [finished] batches: [20.75s] [188it] [ 9.06it/s]


name      nDCG   nDCG@10  AP(rel=2)  P(rel=2)@10  nDCG +  nDCG -  \
0           TFIDF  0.123589  0.595818   0.054568     0.546667     NaN     NaN   
1  TFIDF >> VBERT  0.118277  0.479929   0.049198     0.420000     3.0    12.0   

   nDCG p-value  nDCG@10 +  nDCG@10 -  nDCG@10 p-value  AP(rel=2) +  \
0           NaN        NaN        NaN              NaN          NaN   
1       0.11542        4.0       11.0          0.02287          3.0   

   AP(rel=2) -  AP(rel=2) p-value  P(rel=2)@10 +  P(rel=2)@10 -  \
0          NaN                NaN            NaN            NaN   
1         12.0           0.012239            3.0           12.0   

   P(rel=2)@10 p-value  
0                  NaN  
1             0.024362

Comme nous le constatons, bien que le modèle soit pré-entraîné, il n'obtient pas de très bons résultats. Cela est dû au fait qu'il n'est pas réglé pour la tâche de classement par pertinence.

Cependant, nous pouvons entraîner le modèle pour le classement (comme indiqué ci-dessus pour KNRM).

## monoT5

Le modèle [monoT5](https://arxiv.org/abs/2003.06713) évalue les documents à l'aide d'un modèle de langage causal. Voyons comment cette approche fonctionne sur TREC COVID.

La classe `MonoT5ReRanker` de `pyterrier_t5` charge automatiquement une version du classeur monoT5 qui est entraînée sur le jeu de données MS MARCO passage.

In [16]:
from pyterrier_t5 import MonoT5ReRanker
monoT5 = MonoT5ReRanker(text_field='abstract')

/usr/local/lib/python3.9/dist-packages/transformers/models/t5/tokenization_t5.py:163: FutureWarning: This tokenizer was incorrectly instantiated with a model max length of 512 which will be corrected in Transformers v5.
For now, this behavior is kept to avoid breaking backwards compatibility when padding/encoding with `truncation is True`.
- Be aware that you SHOULD NOT rely on t5-base automatically truncating your input to 512 when padding/encoding.
- If you want to encode/pad to sequences longer than 512 you can either instantiate this tokenizer with `model_max_length` or pass `max_length` when encoding/padding.
- To avoid this warning, please instantiate this tokenizer with `model_max_length` set to your preferred value.
  warnings.warn(


In [17]:
pipeline = (tfidf >> get_text >> monoT5)
pt.Experiment(
    [tfidf, pipeline],
    test_topics,
    qrels,
    names=['TFIDF', 'TFIDF >> T5'],
    eval_metrics=[AP(rel=2), nDCG, nDCG@10, P(rel=2)@10, "mrt"]
)

monoT5:   0%|          | 0/188 s<?, ?batches/s]

Token indices sequence length is longer than the specified maximum sequence length for this model (573 > 512). Running this sequence through the model will result in indexing errors


name  AP(rel=2)      nDCG   nDCG@10  P(rel=2)@10          mrt
0        TFIDF   0.054568  0.123589  0.595818     0.546667    52.372967
1  TFIDF >> T5   0.058684  0.128119  0.685513     0.600000  1701.459983

Comme on pouvait s'y attendre, les résultats sont bien meilleurs en termes de NDCG@10 (0.5958 vs 0.6855).

# Tâche pratique

Comme pour le TP precedent, utilisez les modèles implémentés pour cord19 dans une tâche de question-réponse. Dans ce contexte, les requêtes sont de questions et les documents sont des documents qui pourraient contenir la réponse. Notez que vous devez refaire l'indexation ainsi que les autres étapes étudiées dans ce TP. Vous pouvez le télécharger le dataset en utilisant les lines de code ci-dessous.

In [18]:
fiqa = {}
fiqa['train'] = pt.datasets.get_dataset('irds:beir/fiqa/train')
fiqa['valid'] = pt.datasets.get_dataset('irds:beir/fiqa/dev')
fiqa['test'] = pt.datasets.get_dataset('irds:beir/fiqa/test')

test_topics = fiqa['test'].get_topics(variant='text')
test_qrels = fiqa['test'].get_qrels()

train_topics = fiqa['train'].get_topics(variant='text')
train_qrels = fiqa['train'].get_qrels()

valid_topics = fiqa['valid'].get_topics(variant='text')
valid_qrels = fiqa['valid'].get_qrels()

[INFO] [starting] opening zip file
[INFO] If you have a local copy of https://public.ukp.informatik.tu-darmstadt.de/thakur/BEIR/datasets/fiqa.zip, you can symlink it here to avoid downloading it again: /root/.ir_datasets/downloads/17918ed23cd04fb15047f73e6c3bd9d9
[INFO] [starting] https://public.ukp.informatik.tu-darmstadt.de/thakur/BEIR/datasets/fiqa.zip
[INFO] [finished] https://public.ukp.informatik.tu-darmstadt.de/thakur/BEIR/datasets/fiqa.zip: [1.06s] [17.9MB] [16.9MB/s]
[INFO] [finished] opening zip file [1.56s]
[INFO] [starting] opening zip file
[INFO] [finished] opening zip file s]
[INFO] [starting] opening zip file
[INFO] [finished] opening zip file s]
[INFO] [starting] opening zip file
[INFO] [finished] opening zip file s]


In [28]:
!pip install -q --upgrade fastrank lightgbm==3.1.1
!pip install -q python-terrier

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 28.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 24.8 MB/s eta 0:00:00


In [36]:
from sklearn.ensemble import RandomForestRegressor
import lightgbm as lgb
import fastrank
RANK_CUTOFF = 10
SEED=42
ltr_feats1 = (tfidf % RANK_CUTOFF) >> pt.text.get_text(cord19, ["title", "date", "doi"]) >> (
    pt.transformer.IdentityTransformer()
    ** # score of text for query 'coronavirus covid'
    (pt.apply.query(lambda row: 'coronavirus covid') >> tfidf)
    ** # score of title (not originally indexed)
    (pt.text.scorer(body_attr="title", takes='docs', wmodel='TF_IDF') ) 
    ** # date 2020
    (pt.apply.doc_score(lambda row: int("2020" in row["date"])))
    ** # has doi
    (pt.apply.doc_score(lambda row: int( row["doi"] is not None and len(row["doi"]) > 0) ))
    ** # abstract coordinate match
    pt.BatchRetrieve(index_ref, wmodel="CoordinateMatch")
)

In [34]:
def append_baseline(system, baseline, max_results = 1000):
    return (system ^ baseline) % max_results

In [ ]:
train_request = fastrank.TrainRequest.coordinate_ascent()

params = train_request.params
params.init_random = True
params.normalize = True
params.seed = 1234567

ca_pipe = ltr_feats1 >> pt.ltr.apply_learned_model(train_request, form='fastrank')

%time ca_pipe.fit(train_topics, train_qrels)

rf = RandomForestRegressor(n_estimators=400, verbose=1, random_state=42, n_jobs=2)

rf_pipe = ltr_feats1 >> pt.ltr.apply_learned_model(rf, form='regression')


%time rf_pipe.fit(train_topics, train_qrels)
lmart_l = lgb.LGBMRanker(
    task="train",
    silent=False,
    min_data_in_leaf=1,
    min_sum_hessian_in_leaf=1,
    max_bin=255,
    num_leaves=31,
    objective="lambdarank",
    metric="ndcg",
    ndcg_eval_at=[10],
    ndcg_at=[10],
    eval_at=[10],
    learning_rate= .1,
    importance_type="gain",
    num_iterations=100,
    early_stopping_rounds=5
)

lmart_x_pipe = ltr_feats1 >> pt.ltr.apply_learned_model(lmart_l, form="ltr", fit_kwargs={'eval_at':[10]})

%time lmart_x_pipe.fit(train_topics, train_qrels, valid_topics, valid_qrels)
pt.Experiment(
    [tfidf % RANK_CUTOFF, ca_pipe, rf_pipe, lmart_x_pipe],
    test_topics,
    test_qrels, 
    names=["TFIDF",  "TFIDF + CA(6f)", "TFIDF + RF(6f)", "TFIDF + LMart(6f)"],
    baseline=0,
    eval_metrics=["map", "ndcg", "ndcg_cut_10", "mrt"])

lmart_rf= append_baseline(lmart_x_pipe,rf_pipe)

lmart_ca = append_baseline(lmart_x_pipe,ca_pipe)
rf_ca = append_baseline(rf_pipe,ca_pipe)
pt.Experiment(
    [tfidf % RANK_CUTOFF, lmart_ca, lmart_rf, rf_ca],
    test_topics,
    test_qrels, 
    names=["TFIDF",  "TFIDF + CA(6f)", "TFIDF + RF(6f)", "TFIDF + LMart(6f)"],
    baseline=0,
    eval_metrics=["map", "ndcg", "ndcg_cut_10", "mrt"])

21:48:11.117 [main] WARN org.terrier.querying.ApplyTermPipeline - The index has no termpipelines configuration, and no control configuration is found. Defaulting to global termpipelines configuration of 'Stopwords,PorterStemmer'. Set a termpipelines control to remove this warning.


/usr/local/lib/python3.9/dist-packages/pyterrier/ops.py:242: UserWarning: Got number of results different expected from <pyterrier.batchretrieve.TextScorer object at 0x7f52a81bbfd0>, expected 54959 received 55321, feature scores for any missing documents be 0, extraneous documents will be removed
  warn("Got number of results different expected from %s, expected %d received %d, feature scores for any "


Identifiez-vous des avantages par rapport à l'utilisation de Learning to Rank ?